### DISCUSSION TITLE
Useful properties and methods locked behind internal scope

### DISCUSSION THREAD

**Discussion by Ziirf** [QuestPDF/QuestPDF#665] [General]
After I have worked with this library for a while I have found frustration is very useful things being locked behind the internal scope. Like the Image class, having a its size locked behind an internal property, which means that if i want to know the size of an image, I need to instantiate the object twice.

In my use case I also need to convert from millimeters, into points. And there is a nifty little extension method called "ToPoints(this float value, Unit unit)" on float, which is in the library, but again it is locked behind an internal scope. So I had to make a 1:1 copy of this function in my own project.

I can understand why somethings is required to be scoped, but to me it seems there is a lot of stuff that is just locked away, and I don't see any good reason for it. So I would like to know why, or better yet, get access to some of these features.

**Comment by MarcinZiabek** [comment]
Thank you so much for sharing your experience 😀 Let me answer your questions.

> After I have worked with this library for a while I have found frustration is very useful things being locked behind the internal scope. Like the Image class, having a its size locked behind an internal property, which means that if i want to know the size of an image, I need to instantiate the object twice.

Generally speaking, the image resolution does not imply its physical size on the paper. Resolution is used to only determine the target aspect ratio. The final image size is constrained by configuration and other elements. Therefore, those properties are not available as they do not seem to be useful.

> In my use case I also need to convert from millimeters, into points. And there is a nifty little extension method called "ToPoints(this float value, Unit unit)" on float, which is in the library, but again it is locked behind an internal scope. So I had to make a 1:1 copy of this function in my own project.

This extension method is indeed quite useful. However, it also extends a very common `float` type. Making it public may clutter IntelliSense suggestions. Moreover, every QuestPDF FluentAPI method supports the unit type parameter, so it essentially reduces the need for manual conversion.

---

I completely understand your points. The biggest challenge in making certain APIs public is creating a strong contract which makes it difficult to further change and refactor. Therefore, whenever possible, every new API is hidden by default.

I would like to better understand your use-cases and where exactly the aforementioned APIs could become useful. Maybe we will reassess the accessibility level of those functions. 

Wishing you a wonderful day.

**Reply by Ziirf** [reply]
Thanks for your reply, it is a shame that the features wont be available, but you seem to have your reasons, and I do not entirely disagree with. As for my use-case of this library; The company I work for is currently in the process of upgrading to a new pdf generating tool, for various dynamically generated pdf. Which should soon be done, and during the test runs it have decreased the time to generate pdf by 100's of times faster, which is amazing.
Where struggled to generate 100 pages of pdf in 5 minutes before, it blaze through in seconds now, so for that this library is great. - But as we are maintaining some old features, means when need to be able to store the values of sizes in a database, and therefore we convert from milometers to points outside of the actual generating process. - which is where we had to copy paste your ToPoints extention method.
As for the Image, we have to do some calculations of the image size, due to how the database is storing the data, which is unavoidable if we want to support some of the legacy database values are stored etc.. - we just had to load it in through skia too to get the size, it isn't a big deal, would just have been nice if that wasn't necessary .

Can't really share code as it is against company policies, sadly.
